### DISCUSSION TITLE
Assigning a USB device to an instance.

### DISCUSSION THREAD

**Discussion by roadsta** [awawa-dev/HyperHDR#598] [Q&A]
Hello, I have HyperHDR (version 19) running on a Raspberry Pi Zero 2 W with some APA102 LEDs, it works great and is very responsive.

I'm feeding in two USB inputs; one from a HDMI splitter for a FireTV Stick, the other is AV out from the TV that goes into a converter then into the pi. Both are detected fine in the HyperHDR Web UI, but the issue I am having is that the selected device does not save to an instance. The fact that its a USB is the input saves, but the selected specific device doesn't save with the instance.

Is there something I can do with JSON API to change the device? Or would it be possible to amend the code so that the instance also saves the actual device.

Anyones input would be greatly appreciated.

**Comment by awawa-dev** [comment]
Hi
Only HyperHDR windows with the name of the instance in the headers allow you to adjust the settings to the selected instance.
For example, here you have an instance named "Philips Hue" (on a red background in the header next to "instance USB capture"). So you can enable/disable the grabber per instance, but the USB grabber settings are global. BTW JSON API methods don't allow you to save settings, the result is transient until next reboot.
![obraz](https://github.com/awawa-dev/HyperHDR/assets/69086569/21dbacfe-08cd-4003-9e9c-37ca7543b11d)

**Reply by roadsta** [reply]
Hey @awawa-dev thanks for the very quick reply! OK thanks for the tip I didnt notice the header part, so the short answer is currently, this isnt possible.

Im fine with it being transient but assuming there isnt a method to do this using the JSON API?

How difficult would it be to make the USB Capture section not global, ie sticky to instances, is this something I could put in a feature request for?

Thanks.

**Reply by awawa-dev** [reply]
For a single USB grabber, support is global as hardware resources cannot be shared independently between instances. Your case with two grabbers is rather unusual and would require changing the entire infrastructure. Perhaps it is possible to add device selection similar to hardware brightness/contrast/saturation support in JSON API, but I'm afraid that's just the beginning of the trouble as different grabbers may not have the same resolution and different settings (probably all other settings would have to be defaulted). A rather unique case, as you understand.

**Reply by roadsta** [reply]
OK thats what I was afraid of, thanks for sharing your feedback though!  Really appreciate it.

Totally get this is fringe case, but perhaps exposing the just the USB device selection to the API as this is the only part I change would be a more realistic feature request? Either way though, from what you are saying it wouldnt be as simple as changing one line of code.

My only alternative is to code something that will go into the web UI and change the device option, which would be hella clunky.

**Reply by awawa-dev** [reply]
For the reasons I already mentioned, I would stay away from implementing it in the API. Another possible solution is to create  2 HyperHDR databases/configuration files (for each of the grabbers), create a trivial service in any technology of your choice, which would stop the HyperHDR service, copy the selected configuration to the HyperHDR home directory and start it again.

**Reply by awawa-dev** [reply]
Fo example Python and much simpler https://github.com/msoap/shell2http can do it. Of course everybody remotely can execute your prepared custom  command so it's not protected.
